### PR TITLE
Add hidden user folder

### PR DESCRIPTION
**Pull request by icewind1991**
this can be used by apps to have apps that are within the users home folder but are hidden from the normal ui bits.

Currently, the "hidden" part is achieved in the dav plugin excluding it from listing and file search excluding results inside it.

I had originally hoped to have the "hidden" logic be more low level instead of having to implement it separately in each api by my attempts for that didn't work out as they would defeat the purpose of the folder in the first place (having it accessible trough normal endpoints by path and fileid)

### CODE REVIEW

File: apps/dav/lib/Connector/Sabre/Directory.php [hunk 1]
```diff
@@ -266,10 +266,14 @@ public function getChildren() {
 			throw new Locked();
 		}
 
+		$hiddenName = 'hidden_' . \OC_Util::getInstanceId();
```
Comment (PVince81) [inline apps/dav/lib/Connector/Sabre/Directory.php:269]: didn't we want it to also start with a dot ?

File: lib/private/Files/Node/Folder.php [hunk 2]
```diff
@@ -256,8 +256,14 @@ public function search($query) {
 		$mountByMountPoint = ['' => $mount];
 
 		if (!$limitToHome) {
+			$hiddenFolder = 'hidden_' . \OC_Util::getInstanceId();
```
Comment (PVince81) [inline lib/private/Files/Node/Folder.php:259]: is there a place where we could expose this as a function to avoid relying on convention only ?

**Comment by icewind1991** [comment]
added a function get generate the folder name and added a dot prefix

**Review by skjnldsv — APPROVED** [review]
A bit of tests would be awesome :see_no_evil:

**Comment by PVince81** [comment]
- [ ] BUG: create a folder "test" in the web UI then rename it to the hidden folder => this should not be possible
- [ ] BUG: rename hidden folder to another name => this should not be possible
- [ ] BUG: Webdav DELETE on hidden folder did not fail, and I can even recreate it manually
- I assume that a PROPFIND on the hidden folder should not be blocked ? it's not currently

there might be more

yeah, would be good to add maybe integration tests for this to confirm that all file operations are denied there or at least that the folder is protected in some way

**Comment by PVince81** [comment]
- [ ] BUG: I shouldn't be able to share the hidden folder with someone else (per API)
- [ ] BUG: while the folder is invisible in the web UI, if I try to create it, it will suddenly reappear with "could not create because it exists already" and suddenly I see it and can access it

**Comment by PVince81** [comment]
- [ ] don't create activity entries for any entries inside the hidden folder ? for example if apps are creating custom folders and files there

File: lib/private/Files/Filesystem.php [hunk 3]
```diff
@@ -897,4 +897,12 @@ public static function getOwner($path) {
 	public static function getETag($path) {
 		return self::$defaultInstance->getETag($path);
 	}
+
+	public static function getHiddenFolderName(): string {
+		return '.hidden_' . \OC_Util::getInstanceId();
+	}
+
+	public static function isPathHidden($path): string {
```
Comment (artonge) [inline lib/private/Files/Filesystem.php:905]:
```suggestion
	public static function isPathHidden($path): bool {
```

File: apps/dav/lib/Connector/Sabre/Directory.php [hunk 4]
```diff
@@ -266,10 +267,14 @@ public function getChildren() {
 			throw new Locked();
 		}
 
+		$hiddenName = Filesystem::getHiddenFolderName();
```
Comment (artonge) [inline apps/dav/lib/Connector/Sabre/Directory.php:270]:
```suggestion
		$hiddenFolderName = Filesystem::getHiddenFolderName();
```

File: lib/private/Files/Node/Folder.php [hunk 5]
```diff
@@ -256,8 +257,14 @@ public function search($query) {
 		$mountByMountPoint = ['' => $mount];
 
 		if (!$limitToHome) {
+			$hiddenFolder = Filesystem::getHiddenFolderName();
 			$mounts = $this->root->getMountsIn($this->path);
 			foreach ($mounts as $mount) {
+				// don't search in mounts inside the hidden folder
+				if (strpos($mount->getMountPoint(), '/' . $hiddenFolder .'/') !== false) {
```
Comment (artonge) [inline lib/private/Files/Node/Folder.php:264]:
```suggestion
				if (Filesystem::isPathHidden($mount->getMountPoint())) {
```

File: lib/private/Files/Node/Root.php [hunk 6]
```diff
@@ -392,6 +394,16 @@ public function getUserFolder($userId) {
 		return $this->userFolderCache->get($userId);
 	}
 
+	public function getHiddenUserFolder($userId) {
+		$userFolder = $this->getUserFolder($userId);
+		$hiddenName = Filesystem::getHiddenFolderName();
```
Comment (artonge) [inline lib/private/Files/Node/Root.php:399]:
```suggestion
		$hiddenFolderName = Filesystem::getHiddenFolderName();
```

File: lib/private/Files/View.php [hunk 7]
```diff
@@ -1661,9 +1661,16 @@ private function searchCommon($method, $args) {
 				}
 			}
 
+			$hiddenFolder = Filesystem::getHiddenFolderName();
 			$mounts = Filesystem::getMountManager()->findIn($this->fakeRoot);
 			foreach ($mounts as $mount) {
 				$mountPoint = $mount->getMountPoint();
+
+				// don't search in mounts inside the hidden folder
+				if (strpos($mountPoint, '/' . $hiddenFolder .'/') !== false) {
```
Comment (artonge) [inline lib/private/Files/View.php:1670]:
```suggestion
				if (Filesystem::isPathHidden($mountPoint)) {
```

**Review by artonge — CHANGES_REQUESTED** [review]
From reading the code:

- What if the user creates a folder with the same name as the hidden folder but in a subfolder.

```
/
/.hidden_azeudhazud    <-- real hidden folder
/sub_folder
/sub_folder/.hidden_azeudhazud    <-- user creates this folder that should maybe not be hidden.
```

If I get this right, the created folder will be considered as hidden by `FileSystem::isPathHidden`.

Is this intended behaviour?

It makes sense if `userA` share his root folder with `userB`


```
/    <-- userB's root folder
/.hidden_azeudhazud    <-- userB's hidden folder
/usera_shared_root_folder    <-- what the name says
/usera_shared_root_folder/.hidden_azeudhazud    <-- userA's hidden folder that should be hidden
```

**Pull request closed without merging**